api/tests/opentrons/hardware_control/test_ot3_api.py

```diff
@@ -1671,3 +1683,51 @@ async def test_tip_presence_disabled_ninety_six_channel(
         await ot3_hardware.pick_up_tip(OT3Mount.LEFT, 60)
 
         tip_present.assert_not_called()
+
+
+@pytest.mark.parametrize(
+    argnames=["old_state", "new_state", "should_trigger"],
+    argvalues=[
+        [EstopState.DISENGAGED, EstopState.NOT_PRESENT, False],
+        [EstopState.DISENGAGED, EstopState.PHYSICALLY_ENGAGED, True],
+        [EstopState.LOGICALLY_ENGAGED, EstopState.PHYSICALLY_ENGAGED, True],
+        [EstopState.NOT_PRESENT, EstopState.PHYSICALLY_ENGAGED, True],
+        [EstopState.PHYSICALLY_ENGAGED, EstopState.LOGICALLY_ENGAGED, False],
+        [EstopState.PHYSICALLY_ENGAGED, EstopState.PHYSICALLY_ENGAGED, False],
+    ],
+)
+async def test_estop_event_deactivate_module(
+    ot3_hardware: ThreadManager[OT3API],
+    decoy: Decoy,
+    old_state: EstopState,
+    new_state: EstopState,
+    should_trigger: bool,
+) -> None:
+    """Test the helper to deactivate modules."""
+    api = ot3_hardware.wrapped()
+    api._backend.module_controls = decoy.mock(cls=AttachedModulesControl)
+    tc = decoy.mock(cls=Thermocycler)
+    hs = decoy.mock(cls=HeaterShaker)
+    md = decoy.mock(cls=MagDeck)
+    td = decoy.mock(cls=TempDeck)
+
+    decoy.when(hs.speed_status).then_return(SpeedStatus.HOLDING)
+
+    decoy.when(api._backend.module_controls.available_modules).then_return(
+        [tc, hs, md, td]
+    )
+
+    estop_event = EstopStateNotification(old_state=old_state, new_state=new_state)
+
+    api._update_estop_state(estop_event)
+
+    await asyncio.sleep(0.1)
```

Comment:
I think you can do this instead if you don't want to use `asyncio.sleep`? 
```suggestion
    await asyncio.get_running_loop().run_in_executor(None, api._update_estop_state, estop_event)
```

Reply:
Wait actually this passes only ~50% of the time... I don't it

Reply:
Yeah I tried using `asyncio.all_tasks` to check for when the shutoff tasks all completed but that didn't seem to actually pick up any of the tasks and it would fail most of the time. The actual tasks finish pretty much instantly so this really just has to release the event loop, but I'm not sure why they won't show up 🤷 

I could set up the callback to return all of the futures and just ignore that return value when it gets called by the estop state machine.........